### ISSUE TITLE
Vite 6 compatibility issue `resolve.conditions`

### ISSUE THREAD

**Issue by thebanjomatic** [vitest-dev/vitest#7070]
### Describe the bug

When using a browser-based evironment (happy-dom, jsdom, ~browser~), vitest still passes `conditions: ['node']` in the resolve config.

When using Vite 5, this doesn't really matter because it would infer that we were targetting a web environment and it would resolve in that context anyway. I'm not 100% sure on how this worked before, but `tryNodeResolve` was being passed a `targetWeb` boolean and the returned ID will match the browser / default import conditions instead of the node condition.

With Vite 6 a lot of this default environment magic was stripped out as part of the Environments api refactor, and now it's just using the resolve.conditions from the corresponding client/ssr configs. This results in a change in behavior where Vitest now uses the node condition for browser environments when using Vite 6.

Using the node condition instead of browser can result in test failures due to differences in behavior between browser and node implementations of a library, but I've also run into issues where esbuild as part of deps optimization generates invalid code where the module itself is "babel compatible" but tries to deconstruct the named exports from `import_mod1.default` instead of `import_mod1`.

These issues reproduce in the latest Vitest `2.1.8` using a resolution for Vite `6.0.3`, as well as Vitest `3.0.0-beta.2`.

### Reproduction

I'm still working on creating a reproducer that I can share and should be able to do so by the end of the day. 

In terms of the actual code changes, I think these two locations can be changed to use:
```ts
conditions: [testConfig.environment === 'node' ? 'node' : 'browser']
```

https://github.com/vitest-dev/vitest/blob/4e60333dc7235704f96314c34ca510e3901fe61f/packages/vitest/src/node/plugins/index.ts#L88-L94

https://github.com/vitest-dev/vitest/blob/4e60333dc7235704f96314c34ca510e3901fe61f/packages/vitest/src/node/plugins/workspace.ts#L67-L73

But it's also a little unclear to me if we should be clearing out the `mainFields: [],` for the browser-based environments also.

### System Info

```shell
System:
    OS: Windows 11 10.0.22631
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i7-12800H
    Memory: 11.40 GB / 31.64 GB
  Binaries:
    Node: 20.18.0 - ~\AppData\Local\fnm_multishells\18680_1733347780000\node.EXE
    Yarn: 4.5.3 - ~\AppData\Local\fnm_multishells\18680_1733347780000\yarn.CMD
    npm: 10.9.1 - D:\source\delete-me\test-vite6\node_modules\.bin\npm.CMD
    bun: 1.1.29 - ~\.bun\bin\bun.EXE
  Browsers:
    Edge: Chromium (127.0.2651.74)
    Internet Explorer: 11.0.22621.3527
  npmPackages:
    vite: ^6.0.0 => 6.0.3
```

### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
There is no compatibility issue here. `happy-dom` and `jsdom` tests are running in Node.js with browser globals. There is no widespread decision on how the `browser` condition should be treated - some libraries expect imports to follow bundlers logic (don't use extensions) which breaks in Vitest when externalized, some just expect globals to be present. Vitest added these conditions for a reason and we are keeping them. The same can be said about the `mainFields` treatment - since Node.js can't resolve those, it's better to remove them altogether to avoid module hazard if the module is externalized.

> When using a browser-based evironment (happy-dom, jsdom, browser)

The browser mode doesn't modify resolve conditions.

**Comment by sheremet-va** [comment]
> This results in a change in behavior where Vitest now uses the node condition for browser environments when using Vite 6.

I missed this part - we might need to inject default conditions now in the config, but we shouldn't remove the `node` condition.

**Comment by sheremet-va** [comment]
> Reproducer: [stackblitz.com/edit/vitest-dev-vitest-xzepei6n?file=package.json&view=editor](https://stackblitz.com/edit/vitest-dev-vitest-xzepei6n?file=package.json&view=editor)

I just want to note that "browser" here is the "happy-dom" environment, not the actual browser, the browser mode in Vitest uses a separate Vite server.

**Comment by thebanjomatic** [comment]
Yes that's correct. I probably should have been more clear in the naming of the project. I think that's fine and I mainly want things to continue working the same as they do in Vite 5. Since there is no intention of changing away from `conditions: ["node"]`, I have updated the reproducer to reflect that and to instead only focus on the failure case:

https://stackblitz.com/edit/vitest-dev-vitest-xzepei6n?file=package.json

Vite 5:
![image](https://github.com/user-attachments/assets/ee298cd3-39be-4490-a787-6931852887b0)

Vite 6:
![image](https://github.com/user-attachments/assets/aa0e8982-a80a-454f-9019-a0b5c1ce0944)

**Comment by thebanjomatic** [comment]
From testing, it looks like adding back in the default conditions for vite >=6 fixes this issue. I'll submit a PR.

**Comment by hi-ogawa** [comment]
Re-opening as PR is reverted https://github.com/vitest-dev/vitest/pull/7271

**Comment by hi-ogawa** [comment]
In the team meeting, we discussed that excluding `module` from `conditions` is considered as a fix to align with Node better. This wasn't previously possible in Vite 5 as Vite implicitly added `module` condition in Vite's `resolveId` API, but excluding it became finally possible on Vite 6. Also note that we already excluded `module` from `mainFields` for a long time and this is a desired direction.

I haven't checked the repro properly, but if there's an issue with a specific package `tslib` or optimizeDeps, then it's possible that there is a bug on package side or deps optimization resolution.

**Comment by vafanassieff** [comment]
Hey there,

In my monorepo, I used to be able with `vite` 5 and `vitest` 2 to use the `development` condition to not build my packages before running tests.

```
".": {
  "development": "./src/index.ts",
  "types": "./src/index.ts",
  "import": "./dist/index.js"
}
```

When using `vitest 3.0.1` this is not the case anymore and adding the condition in Vite `resolve.conditions` or in `execArgv` does not work.

I think #7271 by fixing the require/import of cjs broked this behavior.

I patched `vitest` locally to add `conditions: ["node", "development"]`

Best

**Comment by kolorfilm** [comment]
> Hey there,
> 
> In my monorepo, I used to be able with `vite` 5 and `vitest` 2 to use the `development` condition to not build my packages before running tests.
> 
> ```
> ".": {
>   "development": "./src/index.ts",
>   "types": "./src/index.ts",
>   "import": "./dist/index.js"
> }
> ```
> 
> When using `vitest 3.0.1` this is not the case anymore and adding the condition in Vite `resolve.conditions` or in `execArgv` does not work.
> 
> I think [#7271](https://github.com/vitest-dev/vitest/pull/7271) by fixing the require/import of cjs broked this behavior.
> 
> I patched `vitest` locally to add `conditions: ["node", "development"]`
> 
> Best

This worked for me. Unfortunately I breaks my local app (tests working) but the app is broken.

```
const vitestConfig = defineVitestConfig({
  resolve: {
    conditions: ['node', 'development'],
  },
  test: {
    setupFiles: './src/setupTests.ts',
    globals: true,
    environment: 'jsdom',
    mockReset: true,
  },
});

export default defineViteConfig(({ command, mode }) => {
    ...
  };

  return mergeConfig(viteConfig, vitestConfig);
});
```

It seems it merges the config. How can I only define these `conditions` for the vitest tests?

**Comment by silverwind** [comment]
Reproduction here: https://github.com/vitest-dev/vitest/issues/7277

I wouldn't classify something that breaks all tests as a "minor bug".

**Comment by thebanjomatic** [comment]
And I guess if this issue isn't appropriate to re-open, should I file another that is more specific to tslib and and the above reproducer: https://stackblitz.com/edit/vitest-dev-vitest-xzepei6n?file=package.json

**Comment by silverwind** [comment]
FWIW, 3.0.4 did fix my issue in https://github.com/vitest-dev/vitest/issues/7277 with the `@neodrag/react` module, so I hope future efforts don't regress that 😆.

**Comment by thebanjomatic** [comment]
After a little more investigating, if I change node_modules/tslib/modules/index.js from:
```js
import tslib from '../tslib.js';
```
to
```js
import * as tslib from '../tslib.js';
```

Then that fixes the problems I'm seeing with vitest on the latest 3.0.4 + vite 6

So it might actually be a simple fix if that's all that is required.
